core/utils/lazy.go

```diff
@@ -0,0 +1,43 @@
+package utils
+
+import (
+	"sync"
+)
+
+type LazyLoad[T any] struct {
+	f     func() (T, error)
+	state T
+	lock  sync.RWMutex
+	once  sync.Once
+}
+
+func NewLazyLoad[T any](f func() (T, error)) *LazyLoad[T] {
+	return &LazyLoad[T]{
+		f: f,
+	}
+}
+
+func (l *LazyLoad[T]) Get() (out T, err error) {
+
+	// fetch only once (or whenever cleared)
+	l.lock.Lock()
+	l.once.Do(func() {
+		l.state, err = l.f()
+	})
+	l.lock.Unlock()
+
+	// if err, clear so next get will retry
+	if err != nil {
+		l.Reset()
+	}
```

Comment:
Could inline the err reset, instead of relocking:
```suggestion
	// if err, clear so next get will retry
	if err != nil {
		l.once = sync.Once{}
	}
	l.lock.Unlock()
```
I started out suggesting this as an optimization, but now I think it could necessary to prevent a race :thinking:

Reply:
good catch! i think you're right since i drop the write lock briefly